Mypy error in certain formula engine

### What happened?

When adding a Power object to a formula engine mypy complains with different error messages

#### Example 1

On checking this following code:

```python
consumer_excess_power_engine = (
    microgrid.producer().power
    + Power.from_watts(-1000)
).build("excess_power")
```

mypy reports:

```
error: Incompatible types in assignment (expression has type "Any | None", variable has type "Sample[Any]")  [assignment]
Found 1 error in 1 file (checked 1 source file)
```

#### Example 2

On checking this following code:

```python
consumer_excess_power_engine = (
    microgrid.producer().power
    + microgrid.consumer().power
    + Power.from_watts(-1000)
).build("excess_power")
```

mypy reports:

```
error: "Sample3Phase[Any]" has no attribute "value"; maybe "value_p1", "value_p2", or "value_p3"?  [attr-defined]
Found 1 error in 1 file (checked 1 source file)
```

#### Example 3

On contrast checking this following, having the Power object in the middle code works:

```python
consumer_excess_power_engine = (
    microgrid.producer().power
    + Power.from_watts(-1000)
    + microgrid.consumer().power
).build("excess_power")
```

#### Example 4

This code doesn't work.

```python
consumer_excess_power_engine = (
    Power.from_watts(-1000)
    + microgrid.producer().power
).build("excess_power")
```

```
find_bug.py:21: error: "Power" has no attribute "build"  [attr-defined]
find_bug.py:23: error: Unsupported operand types for + ("Power" and "FormulaEngine[Power]")  [operator]
Found 2 errors in 1 file (checked 1 source file)
```

### What did you expect instead?

For the examples 1-3 that work mypy shouldn't complain. We may investigate a way to make Example 4 work.

### Affected version(s)

v1.x.x

### Affected part(s)

Data pipeline (part:data-pipeline)

### Extra information

_No response_

This is the example app that I was using

```python
import asyncio

from datetime import timedelta
from frequenz.sdk import microgrid
from frequenz.sdk.actor import ResamplerConfig
from frequenz.sdk.timeseries import Power

async def run() -> None:
    # This points to the default Frequenz microgrid sandbox
    microgrid_host = "microgrid.sandbox.api.frequenz.io"
    microgrid_port = 62060

    # Initialize the microgrid
    await microgrid.initialize(
        microgrid_host,
        microgrid_port,
        ResamplerConfig(resampling_period=timedelta(seconds=1)),
    )

    # negative means feed-in power due to the negative sign convention
    consumer_excess_power_engine = (
        microgrid.producer().power
        + microgrid.consumer().power
        + Power.from_watts(-1000)
    ).build("excess_power") # (2)!
    cons_excess_power_recv = consumer_excess_power_engine.new_receiver()

    async for cons_excess_power in cons_excess_power_recv:
        cons_excess_power = cons_excess_power.value
        if cons_excess_power is None:
            continue
        print(f"Consumer excess power: {cons_excess_power}")

def main() -> None:
    asyncio.run(run())

if __name__ == "__main__":
    main()
```

and you where right, the bug is not in the shared lines but when accessing the received value.

```
find_bug.py:28: error: "Sample3Phase[Any]" has no attribute "value"; maybe "value_p1", "value_p2", or "value_p3"?  [attr-defined]
Found 1 error in 1 file (checked 1 source file)
```

--- a/RELEASE_NOTES.md
+++ b/RELEASE_NOTES.md
@@ -17,4 +17,4 @@
 
 ## Bug Fixes
 
-<!-- Here goes notable bug fixes that are worth a special mention or explanation -->
+- Fixed a typing issue that occurs in some cases when composing formulas with constants.

--- a/src/frequenz/sdk/timeseries/formula_engine/_formula_engine.py
+++ b/src/frequenz/sdk/timeseries/formula_engine/_formula_engine.py
@@ -842,12 +842,12 @@ class _BaseHOFormulaBuilder(ABC, Generic[QuantityT]):
 
     @overload
     def __add__(
-        self, other: _CompositionType1Phase
+        self, other: _CompositionType1Phase | QuantityT
     ) -> HigherOrderFormulaBuilder[QuantityT]: ...
 
     @overload
     def __add__(
-        self, other: _CompositionType3Phase | QuantityT
+        self, other: _CompositionType3Phase
     ) -> HigherOrderFormulaBuilder3Phase[QuantityT]: ...
 
     def __add__(
@@ -870,12 +870,12 @@ class _BaseHOFormulaBuilder(ABC, Generic[QuantityT]):
 
     @overload
     def __sub__(
-        self, other: _CompositionType1Phase
+        self, other: _CompositionType1Phase | QuantityT
     ) -> HigherOrderFormulaBuilder[QuantityT]: ...
 
     @overload
     def __sub__(
-        self, other: _CompositionType3Phase | QuantityT
+        self, other: _CompositionType3Phase
     ) -> HigherOrderFormulaBuilder3Phase[QuantityT]: ...
 
     def __sub__(
@@ -899,12 +899,12 @@ class _BaseHOFormulaBuilder(ABC, Generic[QuantityT]):
 
     @overload
     def __mul__(
-        self, other: _CompositionType1Phase
+        self, other: _CompositionType1Phase | float
     ) -> HigherOrderFormulaBuilder[QuantityT]: ...
 
     @overload
     def __mul__(
-        self, other: _CompositionType3Phase | float
+        self, other: _CompositionType3Phase
     ) -> HigherOrderFormulaBuilder3Phase[QuantityT]: ...
 
     def __mul__(
@@ -928,12 +928,12 @@ class _BaseHOFormulaBuilder(ABC, Generic[QuantityT]):
 
     @overload
     def __truediv__(
-        self, other: _CompositionType1Phase
+        self, other: _CompositionType1Phase | float
     ) -> HigherOrderFormulaBuilder[QuantityT]: ...
 
     @overload
     def __truediv__(
-        self, other: _CompositionType3Phase | float
+        self, other: _CompositionType3Phase
     ) -> HigherOrderFormulaBuilder3Phase[QuantityT]: ...
 
     def __truediv__(
@@ -957,12 +957,12 @@ class _BaseHOFormulaBuilder(ABC, Generic[QuantityT]):
 
     @overload
     def max(
-        self, other: _CompositionType1Phase
+        self, other: _CompositionType1Phase | QuantityT
     ) -> HigherOrderFormulaBuilder[QuantityT]: ...
 
     @overload
     def max(
-        self, other: _CompositionType3Phase | QuantityT
+        self, other: _CompositionType3Phase
     ) -> HigherOrderFormulaBuilder3Phase[QuantityT]: ...
 
     def max(
@@ -985,12 +985,12 @@ class _BaseHOFormulaBuilder(ABC, Generic[QuantityT]):
 
     @overload
     def min(
-        self, other: _CompositionType1Phase
+        self, other: _CompositionType1Phase | QuantityT
     ) -> HigherOrderFormulaBuilder[QuantityT]: ...
 
     @overload
     def min(
-        self, other: _CompositionType3Phase | QuantityT
+        self, other: _CompositionType3Phase
     ) -> HigherOrderFormulaBuilder3Phase[QuantityT]: ...
 
     def min(

--- a/tests/timeseries/_formula_engine/test_formula_composition.py
+++ b/tests/timeseries/_formula_engine/test_formula_composition.py
@@ -12,7 +12,7 @@ from frequenz.client.microgrid import ComponentMetricId
 from pytest_mock import MockerFixture
 
 from frequenz.sdk import microgrid
-from frequenz.sdk.timeseries._quantities import Power
+from frequenz.sdk.timeseries import Power, Sample
 
 from ..mock_microgrid import MockMicrogrid
 from .utils import get_resampled_stream
@@ -303,7 +303,10 @@ class TestFormulaComposition:
             max_pow = await engine_max_rx.receive()
             assert max_pow and max_pow.value and max_pow.value.isclose(Power.zero())
 
-    async def test_formula_composition_constant(self, mocker: MockerFixture) -> None:
+    async def test_formula_composition_constant(  # pylint: disable=too-many-locals
+        self,
+        mocker: MockerFixture,
+    ) -> None:
         """Test the composition of formulas with constant values."""
         async with (
             MockMicrogrid(grid_meter=True, mocker=mocker) as mockgrid,
@@ -334,10 +337,21 @@ class TestFormulaComposition:
                 engine_div._stop  # pylint: disable=protected-access
             )
 
+            engine_composite = (
+                (
+                    (grid.power + Power.from_watts(50.0)) / 2.0
+                    + grid.power
+                    - Power.from_watts(20.0)
+                )
+                * 2.0
+            ).build("grid_power_composite")
+
             await mockgrid.mock_resampler.send_meter_power([100.0])
 
             # Test addition
-            grid_power_addition = await engine_add.new_receiver().receive()
+            grid_power_addition: Sample[Power] = (
+                await engine_add.new_receiver().receive()
+            )
             assert grid_power_addition.value is not None
             assert math.isclose(
                 grid_power_addition.value.as_watts(),
@@ -345,7 +359,9 @@ class TestFormulaComposition:
             )
 
             # Test subtraction
-            grid_power_subtraction = await engine_sub.new_receiver().receive()
+            grid_power_subtraction: Sample[Power] = (
+                await engine_sub.new_receiver().receive()
+            )
             assert grid_power_subtraction.value is not None
             assert math.isclose(
                 grid_power_subtraction.value.as_watts(),
@@ -353,7 +369,9 @@ class TestFormulaComposition:
             )
 
             # Test multiplication
-            grid_power_multiplication = await engine_mul.new_receiver().receive()
+            grid_power_multiplication: Sample[Power] = (
+                await engine_mul.new_receiver().receive()
+            )
             assert grid_power_multiplication.value is not None
             assert math.isclose(
                 grid_power_multiplication.value.as_watts(),
@@ -361,12 +379,21 @@ class TestFormulaComposition:
             )
 
             # Test division
-            grid_power_division = await engine_div.new_receiver().receive()
+            grid_power_division: Sample[Power] = (
+                await engine_div.new_receiver().receive()
+            )
             assert grid_power_division.value is not None
             assert math.isclose(
                 grid_power_division.value.as_watts(),
                 50.0,
             )
+
+            # Test composite formula
+            grid_power_composite: Sample[Power] = (
+                await engine_composite.new_receiver().receive()
+            )
+            assert grid_power_composite.value is not None
+            assert math.isclose(grid_power_composite.value.as_watts(), 310.0)
 
             # Test multiplication with a Quantity
             with pytest.raises(RuntimeError):

